src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -0,0 +1,982 @@
+           IEnumerable<SyntaxNode> nodesToAnalyze,
+           IDictionary<TLanguageKindEnum, ImmutableArray<SyntaxNodeAnalyzerAction<TLanguageKindEnum>>> nodeActionsByKind,
+           DiagnosticAnalyzer analyzer,
+           SemanticModel model,
+           Func<SyntaxNode, TLanguageKindEnum> getKind,
+           TextSpan filterSpan,
+           SyntaxReference declaration,
+           AnalysisScope analysisScope,
+           AnalysisState analysisStateOpt)
+           where TLanguageKindEnum : struct
+        {
+            DeclarationAnalyzerStateData analyzerStateOpt = null;
+
+            try
+            {
+                if (TryStartAnalyzingSyntaxRefence(declaration, analyzer, analysisScope, analysisStateOpt, out analyzerStateOpt))
+                {
+                    ExecuteSyntaxNodeActionsCore(nodesToAnalyze, nodeActionsByKind, analyzer, model, getKind, filterSpan, analyzerStateOpt);
+                }
+            }
+            finally
+            {
+                analyzerStateOpt?.ResetToReadyState();
+            }
+        }
+
+        private void ExecuteSyntaxNodeActionsCore<TLanguageKindEnum>(
+            IEnumerable<SyntaxNode> nodesToAnalyze,
+            IDictionary<TLanguageKindEnum, ImmutableArray<SyntaxNodeAnalyzerAction<TLanguageKindEnum>>> nodeActionsByKind,
+            DiagnosticAnalyzer analyzer,
+            SemanticModel model,
+            Func<SyntaxNode, TLanguageKindEnum> getKind,
+            TextSpan filterSpan,
+            SyntaxNodeAnalyzerStateData analyzerStateOpt)
+            where TLanguageKindEnum : struct
+        {
+            var addDiagnostic = GetAddDiagnostic(model.SyntaxTree, filterSpan, analyzer, isSyntaxDiagnostic: false);
+            ExecuteSyntaxNodeActions(nodesToAnalyze, nodeActionsByKind, model, getKind, addDiagnostic, analyzerStateOpt);
+        }
+
+        private void ExecuteSyntaxNodeActions<TLanguageKindEnum>(
+            IEnumerable<SyntaxNode> nodesToAnalyze,
+            IDictionary<TLanguageKindEnum, ImmutableArray<SyntaxNodeAnalyzerAction<TLanguageKindEnum>>> nodeActionsByKind,
+            SemanticModel model,
+            Func<SyntaxNode, TLanguageKindEnum> getKind,
+            Action<Diagnostic> addDiagnostic,
+            SyntaxNodeAnalyzerStateData analyzerStateOpt)
+            where TLanguageKindEnum : struct
+        {
+            Debug.Assert(nodeActionsByKind != null);
+            Debug.Assert(nodeActionsByKind.Any());
+
+            SyntaxNode partiallyProcessedNode = analyzerStateOpt?.CurrentNode;
+            if (partiallyProcessedNode != null)
+            {
+                ExecuteSyntaxNodeActions(partiallyProcessedNode, nodeActionsByKind, model, getKind, addDiagnostic, analyzerStateOpt);
+            }
+
+            foreach (var child in nodesToAnalyze)
+            {
+                if (ShouldExecuteNode(analyzerStateOpt, child))
+                {
+                    SetCurrentNode(analyzerStateOpt, child);
```

Comment:
Yes, state is a mutable entity tracking current partial analysis of events/symbols/declarations per-analyzer. It is guaranteed to be handed out to a single thread, any other thread attempting to analyze the same entity for same analyzer bails out, so changes to it need not be thread safe.